packages/protocol/contracts/governance/DowntimeSlasherIntervals.sol

```diff
@@ -0,0 +1,409 @@
+      1
+    );
+    uint256 signerIndex = startSignerIndex;
+    uint256 lastIntervalEnd = startIntervals[0];
+    for (uint256 i = 0; i < startIntervals.length; i += 1) {
+      require(
+        intervalProofAlreadyCalculated(startIntervals[i], endIntervals[i]),
+        "Interval has not been calculated"
+      );
+      require(
+        lastIntervalEnd >= startIntervals[i].sub(1) && lastIntervalEnd <= endIntervals[i],
+        "At least one endInterval is not in te boundaries of the next Interval"
+      );
+      if (startIntervals[i] >= epochChangeBlock) {
+        signerIndex = endSignerIndex;
+      }
+      if (!wasDownUsingIntervalProof(startIntervals[i], endIntervals[i], signerIndex)) {
+        return false;
+      }
+      lastIntervalEnd = endIntervals[i];
+      // Could be covered with less intervals
+      if (lastIntervalEnd >= endBlock) {
+        return true;
+      }
+    }
+
+    require(
+      lastIntervalEnd >= endBlock,
+      "The Intervals are not covering the SlashableDowntime window"
+    );
+
+    return true;
+  }
+
+  /**
+   * @notice Returns the end block for the interval required for slashing.
+   */
+  function getEndBlockForSlashing(uint256 startBlock) internal view returns (uint256) {
+    return startBlock.add(slashableDowntime).sub(1);
+  }
+
+  /**
+   * @notice Add to the validator a new SlashedInterval if the validator was not already
+   * slashed by a previous interval that contains al least a block of the new interval.
+   * If the oncePerEpoch flag is set, won't allow two slashes in the same epoch for the same
+   * validator 
+   */
+  function addNewSlashIntervalToValidator(address validator, uint256 startBlock) internal {
+    uint256 endBlock = getEndBlockForSlashing(startBlock);
+    uint256 startEpoch = getEpochNumberOfBlock(startBlock);
+    uint256 endEpoch = getEpochNumberOfBlock(endBlock);
+
+    SlashedInterval[] storage intervals = epochSlashedIntervals[validator][startEpoch];
+    // The oncePerEpoch=true validation
+    require(!oncePerEpoch || intervals.length < 1, "Already slashed in that epoch");
+
+    for (uint256 i = 0; i < intervals.length; i = i.add(1)) {
+      require(
+        intervals[i].endBlock < startBlock || intervals[i].startBlock > endBlock,
+        "Slash shares blocks with another slash"
+      );
+    }
+
+    if (startEpoch != endEpoch) {
+      intervals = epochSlashedIntervals[validator][endEpoch];
+      for (uint256 i = 0; i < intervals.length; i = i.add(1)) {
+        require(
+          intervals[i].endBlock < startBlock || intervals[i].startBlock > endBlock,
+          "Slash shares blocks with another slash"
+        );
+      }
+    }
+    epochSlashedIntervals[validator][startEpoch].push(
+      SlashedInterval({ startBlock: startBlock, endBlock: endBlock })
+    );
+  }
+
+  /**
+   * @notice Requires that `wasDownForIntervals` returns true and that the account corresponding
```

Comment:
Comments this long suggest the code is too complicated or the comment is unnecessarily verbose.

In this case, I think it's the latter.

Reply:
I removed a few, but let me know what do you think (is still big)